src/EditorFeatures/Core/Tagging/AbstractAsynchronousTaggerProvider.TagSource_ProduceTags.cs

```diff
@@ -129,35 +93,51 @@ private void RemoveTagsThatIntersectEdit(TextContentChangedEventArgs e)
             if (e.Changes.Count == 0)
                 return;
 
-            var buffer = e.After.TextBuffer;
-            if (!this.CachedTagTrees.TryGetValue(buffer, out var treeForBuffer))
-                return;
-
             var snapshot = e.After;
+            var buffer = snapshot.TextBuffer;
 
-            var tagsToRemove = e.Changes.SelectMany(c => treeForBuffer.GetIntersectingSpans(new SnapshotSpan(snapshot, c.NewSpan)));
-            if (!tagsToRemove.Any())
-                return;
+            // Everything we're passing in here is synchronous.  So we can assert that this must complete synchronously
+            // as well.
+            var (oldTagTrees, newTagTrees) = CompareAndSwapTagTreesAsync(
+                oldTagTrees =>
+                {
+                    if (oldTagTrees.TryGetValue(buffer, out var treeForBuffer))
+                    {
+                        var tagsToRemove = e.Changes.SelectMany(c => treeForBuffer.GetIntersectingSpans(new SnapshotSpan(snapshot, c.NewSpan)));
+                        if (tagsToRemove.Any())
+                        {
+                            var allTags = treeForBuffer.GetSpans(e.After).ToList();
+                            var newTagTree = new TagSpanIntervalTree<TTag>(
+                                buffer,
+                                treeForBuffer.SpanTrackingMode,
+                                allTags.Except(tagsToRemove, comparer: this));
+                            return new(oldTagTrees.SetItem(buffer, newTagTree));
+                        }
+                    }
 
-            var allTags = treeForBuffer.GetSpans(e.After).ToList();
-            var newTagTree = new TagSpanIntervalTree<TTag>(
-                buffer,
-                treeForBuffer.SpanTrackingMode,
-                allTags.Except(tagsToRemove, comparer: this));
+                    // return oldTagTrees to indicate nothing changed.
+                    return new(oldTagTrees);
+                }, _dataSource.ThreadingContext.DisposalToken).VerifyCompleted();
```

Comment:
Do we not have a cancellation token for buffer is closed? Otherwise are we still processing things after documents have been closed?

Reply:
good point. this may be incorrect.